EKF/mag_fusion.cpp

```diff
@@ -601,21 +601,14 @@ void Ekf::updateQuaternion(const float innovation, const float variance, const f
 		float tmp = 0.0f;
 
 		for (uint8_t col = 0; col <= 3; col++) {
-			tmp += P(row,col) * yaw_jacobian(col);
+			tmp += P(row,col) * yaw_jacobian.atCompressedIndex(col);
```

Comment:
The `atCompressedIndex` is mixing sparse and dense addressing. It technically works _here_, but only because the indices are [0,1,2,3]. Better to not get into this pattern. I see 2 options:

1. Express this nested loop as a matrix operation instead. I think that is the direction we should go for this kind of thing, and add any needed matrix-style operations to the SparseVector class. This would be something like `_heading_innov_var += yaw_jacobian.transpose() * P * yaw_jacobian;` , although maybe we want to make a helper function called something like `mult_btAb(Matrix A, SparseVector b)` to avoid re-implementing Eigen's entire compile-time temporary system.

2. If it really can't be a matrix operation (or we can't implement that operation for some reason), then we need to correctly translate between sparse and dense indices. The loops should iterate over the compressed indices, eg: `for (unsigned compressed_row = 0; compressed_row < yaw_jacobian.non_zeros(); compressed_row++)`, and accesses to P need to be translated into non-compressed indices:
```suggestion
			tmp += P(yaw_jacobian.index(compressed_row), yaw_jacobian.index(compressed_col)) * yaw_jacobian.atCompressedIndex(compressed_col);
```

Reply:
Note that option 2 is basically what would be inside of the helper function described in option 1.